csrc/transformer/normalize_kernels.cu

```diff
@@ -29,26 +19,36 @@ __global__ void fused_bias_residual_layer_norm(float* vals,
                                                bool preLayerNorm,
                                                bool training,
                                                float* vars,
-                                               float* means)
+                                               float* means,
+                                               int row_stride)
 {
-    constexpr int iteration_stride = row_stride / iterations;
+    int iteration_stride = blockDim.x;
+    int iterations = row_stride / iteration_stride;
 
     cg::thread_block b = cg::this_thread_block();
-    cg::thread_block_tile<32> g = cg::tiled_partition<32>(b);
+    cg::thread_block_tile<WARP_SIZE> g = cg::tiled_partition<WARP_SIZE>(b);
 
     int row = blockIdx.x;
     int id = threadIdx.x;
-    int gid = id / 32;
+    int gid = id / WARP_SIZE;
 
-    float vals_arr[iterations];
-    __shared__ float shr[iteration_stride >> 5];
+    float vals_arr[MAX_REGISTERS];
+    __shared__ float shr[MAX_WARP_NUM];
+
+    residual += (row * row_stride);
+    vals += (row * row_stride);
 
     float sum = 0.f;
 #pragma unroll
     for (int i = 0; i < iterations; i++) {
-        vals_arr[i] = residual[row * row_stride + i * iteration_stride + id];
+        vals_arr[i] = residual[i * iteration_stride + id];
         sum += vals_arr[i];
     }
+    if ((iterations * iteration_stride + id) < row_stride) {

```

Comment:
it may be better to assign "iterations * iteration_stride + id" to a variable, and reuse it instead of repeat this expression.

Reply:
I agree. I will do that. Thanks Elton